Reading geopyspark pyramid layer after persisting to s3 throws Avro Error

I have single band tiffs that i am trying to read from s3 and then writing the pyramid to s3 from inside geopyspark as below: 

`
layer = gps.geotiff.get(layer_type=gps.LayerType.SPATIAL,uri="s3://some.s3.payth/data/")
tiled_raster_layer = layer.tile_to_layout(gps.GlobalLayout(), target_crs=3857)
pyramid = tiled_raster_layer.repartition(100).pyramid()

for tiled_layer in pyramid.levels.values():
    gps.write(uri="s3://some.s3.path/catalog", layer_name="example-layer", tiled_raster_layer=tiled_layer)

tms_server = gps.TMS.build(("s3://some.s3.path/catalog", "example-layer"), 
                           display=yield_color_map)
M.add_layer(TMSRasterData(tms_server), name="example-layer")
`

When I try to read the layer back as above with the tms server I see the following stack.  I have tried using a Multiband Tiff, a SingleBandTiff, as well as a single file and multiple files as input.

`
[ERROR] [09/26/2017 04:29:50.287] [geopyspark-tile-server-akka.actor.default-dispatcher-5] [akka.actor.ActorSystemImpl(geopyspark-tile-server)] Error during processing of request HttpRequest(HttpMethod(GET),http://localhost:61799/tile/16/16045/24116.png,List(Host: localhost:61799, Connection: close, Accept-Encoding: gzip, Timeout-Access: <function1>),HttpEntity.Strict(none/none,ByteString()),HttpProtocol(HTTP/1.1))
org.apache.avro.AvroTypeException: Found geotrellis.raster.ArrayMultibandTile, expecting union. This can be caused by using a type parameter which doesn't match the object being deserialized.
	at geotrellis.spark.io.avro.AvroEncoder$.fromBinary(AvroEncoder.scala:94)
	at geotrellis.spark.io.avro.AvroEncoder$.fromBinary(AvroEncoder.scala:77)
	at geotrellis.spark.io.s3.S3ValueReader$$anon$1.read(S3ValueReader.scala:58)
	at geotrellis.spark.io.Reader$class.apply(Reader.scala:21)
	at geotrellis.spark.io.s3.S3ValueReader$$anon$1.apply(S3ValueReader.scala:39)
	at geopyspark.geotrellis.tms.TileReaders$CatalogTileReader$$anonfun$retrieve$1$$anonfun$5$$anonfun$apply$1.apply(TileReader.scala:59)
	at geopyspark.geotrellis.tms.TileReaders$CatalogTileReader$$anonfun$retrieve$1$$anonfun$5$$anonfun$apply$1.apply(TileReader.scala:59)
	at geopyspark.geotrellis.tms.TileReaders$.geopyspark$geotrellis$tms$TileReaders$$rezoom(TileReader.scala:32)
	at geopyspark.geotrellis.tms.TileReaders$CatalogTileReader$$anonfun$retrieve$1$$anonfun$5.apply(TileReader.scala:59)
	at geopyspark.geotrellis.tms.TileReaders$CatalogTileReader$$anonfun$retrieve$1$$anonfun$5.apply(TileReader.scala:59)
	at scala.util.Try$.apply(Try.scala:192)
	at geopyspark.geotrellis.tms.TileReaders$CatalogTileReader$$anonfun$retrieve$1.apply(TileReader.scala:59)
	at geopyspark.geotrellis.tms.TileReaders$CatalogTileReader$$anonfun$retrieve$1.apply(TileReader.scala:57)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24)
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24)
	at scala.concurrent.impl.ExecutionContextImpl$AdaptedForkJoinTask.exec(ExecutionContextImpl.scala:121)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
`

--- a/geopyspark-backend/geotrellis/src/main/scala/geopyspark/geotrellis/tms/TileReader.scala
+++ b/geopyspark-backend/geotrellis/src/main/scala/geopyspark/geotrellis/tms/TileReader.scala
@@ -15,6 +15,7 @@ import scala.collection.concurrent._
 import scala.collection.JavaConversions._
 import scala.util.{Try, Failure, Success}
 
+
 trait TileReader {
   def startup(): Unit = {}
   def shutdown(): Unit = {}
@@ -44,13 +45,11 @@ object TileReaders {
   private class CatalogTileReader(
     valueReader: ValueReader[LayerId],
     catalog: String,
-    overzooming: Boolean
+    overzooming: Boolean,
+    zoomLevels: Seq[Int],
+    maxZoom: Int
   ) extends TileReader {
-
     private val layers = TrieMap.empty[Int, Reader[SpatialKey, Tile]]
-    private val zoomLevels = 
-      for { LayerId(name, zoom) <- valueReader.attributeStore.layerIds if name == catalog } yield zoom
-    private val maxZoom = zoomLevels.max
 
     def retrieve(zoom: Int, x: Int, y: Int) = {
       Future {
@@ -65,6 +64,37 @@ object TileReaders {
           val key = SpatialKey(x, y)
           val reader = layers.getOrElseUpdate(zoom, valueReader.reader[SpatialKey, Tile](LayerId(catalog, zoom)))
           Try(MultibandTile(reader(key))) match {
+            case Success(tile) => Some(tile)
+            case Failure(_: ValueNotFoundError) => None
+            case Failure(e: Throwable) => throw e
+          }
+        }
+      }
+    }
+  }
+
+  private class CatalogMultibandTileReader(
+    valueReader: ValueReader[LayerId],
+    catalog: String,
+    overzooming: Boolean,
+    zoomLevels: Seq[Int],
+    maxZoom: Int
+  ) extends TileReader {
+    private val layers = TrieMap.empty[Int, Reader[SpatialKey, MultibandTile]]
+
+    def retrieve(zoom: Int, x: Int, y: Int): Future[Option[MultibandTile]] = {
+      Future {
+        if (overzooming && zoom > maxZoom) {
+          val reader = layers.getOrElseUpdate(maxZoom, valueReader.reader[SpatialKey, MultibandTile](LayerId(catalog, maxZoom)))
+          Try(rezoom(zoom, x, y, maxZoom, k => reader(k))) match {
+            case Success(tile) => Some(tile)
+            case Failure(_: ValueNotFoundError) => None
+            case Failure(e: Throwable) => throw e
+          }
+        } else {
+          val key = SpatialKey(x, y)
+          val reader = layers.getOrElseUpdate(zoom, valueReader.reader[SpatialKey, MultibandTile](LayerId(catalog, zoom)))
+          Try(reader(key)) match {
             case Success(tile) => Some(tile)
             case Failure(_: ValueNotFoundError) => None
             case Failure(e: Throwable) => throw e
@@ -148,13 +178,13 @@ object TileReaders {
                   val rdd = levels(maxZoom)
                   val kps = reqs.map{ case QueueRequest(_, x, y, promise) => (SpatialKey(x, y), promise) }
                   val dz = zoom - maxZoom
-                  val remap: SpatialKey => SpatialKey = { 
-                    case SpatialKey(x, y) => SpatialKey((x / math.pow(2, dz)).toInt, (y / math.pow(2, dz)).toInt) 
+                  val remap: SpatialKey => SpatialKey = {
+                    case SpatialKey(x, y) => SpatialKey((x / math.pow(2, dz)).toInt, (y / math.pow(2, dz)).toInt)
                   }
                   val keys = kps.map{ case (key, _) => remap(key) }.toSet
                   val rawTiles = new MultiValueRDDFunctions(rdd).multilookup(keys)
                   val fetch: SpatialKey => MultibandTile = { toFind => rawTiles.find{ case (rddKey, _) => rddKey == toFind }.get._2 }
-                  kps.foreach{ case (key, promise) => 
+                  kps.foreach{ case (key, promise) =>
                     promise success (Try(rezoom(zoom, key._1, key._2, maxZoom, fetch)).toOption)
                   }
                 } else
@@ -209,15 +239,36 @@ object TileReaders {
 
   def createCatalogReader(uriString: String, layerName: String, overzooming: Boolean): TileReader = {
     val uri = new java.net.URI(uriString)
-
     val valueReader = ValueReader(uri)
+    val attributeStore = valueReader.attributeStore
 
-    new CatalogTileReader(valueReader, layerName, overzooming)
+    val ids = attributeStore.layerIds
+    val zoomLevels = for { LayerId(name, zoom) <- ids if name == layerName } yield zoom
+    val maxZoom = zoomLevels.max
+
+    val valueClass = attributeStore.readHeader[LayerHeader](ids.head).valueClass
+
+    valueClass match {
+      case "geotrellis.raster.Tile" =>
+        new CatalogTileReader(
+          valueReader,
+          layerName,
+          overzooming,
+          zoomLevels,
+          maxZoom)
+      case "geotrellis.raster.MultibandTile" =>
+        new CatalogMultibandTileReader(
+          valueReader,
+          layerName,
+          overzooming,
+          zoomLevels,
+          maxZoom)
+    }
   }
 
   def createSpatialRddReader(
     levels: java.util.HashMap[Int, TiledRasterLayer[SpatialKey]],
-    system: ActorSystem, 
+    system: ActorSystem,
     overzooming: Boolean
   ): TileReader = {
     val tiles = levels.mapValues(_.rdd)

